### DISCUSSION TITLE
Using BigInteger inside functions

### DISCUSSION THREAD

**Discussion by Apakottur** [sqlalchemy/sqlalchemy#11929] [Usage Questions] [answered]
I am trying to use the BigInteger type inside `func.least()` and I am getting a type conversion issue.
Specifically, I have a DB column of type BigInteger, which corresponds to my Python type BigInt.

My queries automatically convert between the Python and DB type without me having to do any casting, but this does not seem to work with `func.least`.

I'm using asyncpg, here's the full reproduction code:
```python
import asyncio
from typing import override

import sqlalchemy.orm.decl_api
import sqlalchemy.orm.decl_base
import sqlalchemy.orm.loading
import sqlalchemy.orm.properties
import sqlalchemy.sql.sqltypes
from sqlalchemy import BigInteger, func, insert, select, sql
from sqlalchemy.engine.interfaces import Dialect
from sqlalchemy.ext.asyncio import async_sessionmaker, create_async_engine
from sqlalchemy.orm import DeclarativeBase, Mapped, mapped_column

_HOST = "localhost"
_PORT = 5432
_USERNAME = ""
_PASSWORD = ""
_DB_NAME = "sqla_test"
_META_DB_NAME = "postgres"

_DB_STRING = f"postgresql+asyncpg://{_USERNAME}:{_PASSWORD}@{_HOST}:{_PORT}/{_DB_NAME}"
_META_DB_STRING = f"postgresql+asyncpg://{_USERNAME}:{_PASSWORD}@{_HOST}:{_PORT}/{_META_DB_NAME}"

engine = create_async_engine(_DB_STRING)
meta_engine = create_async_engine(_META_DB_STRING, isolation_level="AUTOCOMMIT")


class BigInt(int):
    # Proprietary type with some custom logic.
    pass


class DbBase(DeclarativeBase):
    pass


class _BigIntIdColumn(sqlalchemy.types.TypeDecorator[BigInt]):
    impl = BigInteger

    cache_ok = True

    @property
    def python_type(self) -> type[BigInt]:
        return BigInt

    @override
    def process_result_value(self, value: int | None, dialect: Dialect) -> BigInt | None:
        if value is None:
            return None
        else:
            return BigInt(value)


class Actor(DbBase):
    __tablename__ = "actor"

    id: Mapped[BigInt] = mapped_column(_BigIntIdColumn, primary_key=True)


async def _recreate_db():
    async with meta_engine.connect() as conn:
        await conn.execute(sql.text(f"DROP DATABASE IF EXISTS {_DB_NAME}"))
        await conn.execute(sql.text(f"CREATE DATABASE {_DB_NAME}"))

    async with engine.connect() as conn:
        await conn.run_sync(DbBase.metadata.create_all)
        await conn.commit()


async def _run():
    await _recreate_db()

    async_session = async_sessionmaker(engine, expire_on_commit=False)

    async with async_session() as session:
        b = BigInt(195471636076429315)

        # Works.
        result = await session.execute(
            insert(Actor)
            .values(
                [
                    {"id": b},
                ]
            )
            .returning(Actor.id),
        )
        assert result.scalars().all() == [b]

        # Works.
        result = await session.execute(select(Actor.id).where(Actor.id == b))
        assert result.scalars().all() == [b]

        # Doesn't work.
        result = await session.execute(select(Actor.id).where(Actor.id == func.least(b, b)))
        assert result.scalars().all() == [b]


if __name__ == "__main__":
    asyncio.run(_run())
```

The last DB query results in:
```sqlalchemy.exc.ProgrammingError: (sqlalchemy.dialects.postgresql.asyncpg.ProgrammingError) <class 'asyncpg.exceptions.UndefinedFunctionError'>: operator does not exist: bigint = text```

How can I modify my code to avoid the error?

**Comment by zzzeek** [comment, marked as answer]
asyncpg is raising an error at the point of prepare(), that is, without any parameters sent, this fails:


```py

prepared_stmt = await self._connection.prepare(
    'SELECT actor.id \nFROM actor \nWHERE actor.id = least($1, $2)'
)


<class 'asyncpg.exceptions.UndefinedFunctionError'>: operator does not exist: bigint = text
HINT:  No operator matches the given name and argument types. You might need to add explicit type casts.


```

so we'd assume that's because `least()` I guess can work with any datatype not just ints, and in this case, it sees no datatype and assumes string.   so it would need casts inside of it:


    SELECT actor.id \nFROM actor \nWHERE actor.id = least($1::BIGINT, $2::BIGINT)

when you make a statement like this:

    result = await session.execute(select(Actor.id).where(Actor.id == b))

The direct comparison of a BigInt `Actor.id` to `b` indicates "b" will be coerced into BigInteger and we see it in the SQL:

```
 SELECT actor.id 
FROM actor 
WHERE actor.id = $1::BIGINT

```

but for this statement, the func container doesn't have any coercion for these, because SQLAlchemy has no automatic coercion for your special `BigInt` datatype:

```
(Pdb) f1 = func.least(123, 456)  # regular ints
(Pdb) f1.clauses.clauses   # detects Integer
[BindParameter('%(140485340798928 least)s', 123, type_=Integer()), BindParameter('%(140485340798784 least)s', 456, type_=Integer())]

(Pdb) f2 = func.least(b, b)  # your special int
(Pdb) f2.clauses.clauses  # detects NullType
[BindParameter('%(140485340802144 least)s', 195471636076429315, type_=NullType()), BindParameter('%(140485340802192 least)s', 195471636076429315, type_=NullType())]

```

so the most immedaite way to get that coercion working is to use `literal()`, at least for the first entry:

```
from sqlalchemy import literal

result = await session.execute(select(Actor.id).where(Actor.id == func.least(literal(b, BigInteger), b)))

```

giving us working SQL:

```
 SELECT actor.id 
FROM actor 
WHERE actor.id = least($1::BIGINT, $2)

```

there is also a hook, that to be quite honest is not at all documented right now and was introduced for the purposes of the PostgreSQL range() type.   We can support documenting this attribute and making it supported API, and it is already tested in our test suite.   you can enable it like this:

```py
class BigInt(int):
    # Proprietary type with some custom logic.

    @property
    def __sa_type_engine__(self):
        return _BigIntIdColumn()

```

or otherwise monkeypatch an `__sa_type_engine__` attribute onto your special type.  that will work also.

**Reply by Apakottur** [reply]
Thanks for the detailed answer.

I added the monkeypatch and it works as expected.
However, once the monkeypatch is deployed on my codebase, some DB queries become up to 50x slower, is that expected?
What's even stranger is that removing the monkeypatch doesn't make the queries fast again, and only recreating the DB does.

Does this monkeypatch lead to a change done on Postgres itself?

I'll try to investigate further and understand what leads to the slowness.

**Reply by Apakottur** [reply]
After running some EXPLAIN ANALYZE I see that with the monkeypatch the affected queries are much slower because of the Postgres JIT feature.
Disabling JIT fixes the problem, but I wonder why the monkeypatch leads to the JIT in the first place?

The affected queries don't even have any of the func.least() stuff I was trying to fix.

**Reply by zzzeek** [reply]
you would want to turn on SQL logging and see if the patch is causing there to be more "::BIGINT" casts than before.

that's literally the only thing that this change can affect.
